test/gtest/ucp/test_ucp_am.cc

```diff
@@ -925,22 +923,56 @@ UCS_TEST_P(test_ucp_am_nbx_send_flag, eager, "RNDV_THRESH=128")
 UCP_INSTANTIATE_TEST_CASE(test_ucp_am_nbx_send_flag)
 
 
-class test_ucp_am_nbx_reply : public test_ucp_am_nbx {
+template <class T>
+class test_ucp_am_nbx_reply_var : public T {
 public:
     static void get_test_variants(variant_vec_t &variants)
     {
-        add_variant_values(variants, test_ucp_am_base::get_test_variants, 0);
-        add_variant_values(variants, test_ucp_am_base::get_test_variants,
-                           UCP_AM_SEND_FLAG_REPLY, "reply");
+        T::add_variant_values(variants, T::get_test_variants, 0);
+        T::add_variant_values(variants, T::get_test_variants,
+                              UCP_AM_SEND_FLAG_REPLY, "reply");
     }
 
 protected:
     virtual unsigned get_send_flag() const
     {
-        return get_variant_value(1);
+        return T::get_variant_value(T::get_num_variants());
+    }
+
+    virtual unsigned int get_num_variants() const
+    {
+        return T::get_num_variants() + 1;
+    }
+};
+
+
+template <class T>
+class test_ucp_am_nbx_prereg_var : public T {
+public:
+    static void get_test_variants(variant_vec_t &variants)
+    {
+        T::add_variant_values(variants, T::get_test_variants, 0, "");
+        T::add_variant_values(variants, T::get_test_variants, 1, "prereg");
+    }
+
+protected:
+    virtual bool prereg() const
+    {
+        return T::get_variant_value(T::get_num_variants());
+    }
+
+    virtual unsigned int get_num_variants() const
+    {
+        return T::get_num_variants() + 1;
     }
 };
 
+using test_ucp_am_nbx_reply        = test_ucp_am_nbx_reply_var<test_ucp_am_nbx>;
+using test_ucp_am_nbx_prereg       =
+        test_ucp_am_nbx_prereg_var<test_ucp_am_nbx>;
+using test_ucp_am_nbx_prereg_reply =
+        test_ucp_am_nbx_prereg_var<test_ucp_am_nbx_reply>;
```

Comment:
IMO it's over complicated.. can't we use simple inheritance?

Reply:
I'm trying to get rid of code duplication here.
Because in some classes we need only reply or prereg variant, but eager_datatype need both.

This can be replaced by three classes.

Reply:
maybe have one class but enumerate variants by flags, and not necessarily add all combinations, but only interesting ones?
for example:
add_variant(eager)
add_variant(eager|prereg)
add_variant(eager|reply)
...

Reply:
and we will need to do this in every class that has variants.
it would be many duplicated lines, each class will have add_variant(proto).

I tried this approach, but many extra code was required.